flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/TwoInputStreamTask.java

```diff
@@ -80,7 +91,10 @@ public void init() throws Exception {
 				this.headOperator);
 
 		// make sure that stream tasks report their I/O statistics
-		inputProcessor.setMetricGroup(getEnvironment().getMetricGroup().getIOMetricGroup());
+		inputProcessor.setupMetrics(getEnvironment().getMetricGroup().getIOMetricGroup(), input1WatermarkGauge, input2WatermarkGauge);
+
+		headOperator.getMetricGroup().gauge(MetricNames.IO_CURRENT_INPUT_1_WATERMARK, input1WatermarkGauge);
```

Comment:
The headoperator is never touched in the OperatorChain; we only setup chained operators.

Reply:
i think

Reply:
oh wait. No this is different. For TwoINputStreamTasks we register the input1/2 watermark metrics here, but the common min watermark metric in the operator chain.

Reply:
so for two-input operators you will get three metrics: "input1", "input2", and "input" where the last is the combined min, right?

Reply:
yes

Reply:
however, the separate input1/input2 metrics ignore the `WatermarkMetricPreference`; as in they are never overridden by the output metric. (because that is difficult to do in the OperatorChain since there we have no concept of TwoInputOperators....)

Reply:
I'm wondering if this reusing of metrics for inputs is really a good idea... 😕

Reply:
you're thinking: no special magic for input's outputs, no preferences. It's just a fact that sources will only have an output watermark and no input watermark?

Reply:
yup. sources don't have input watermark metrics, sinks don't have output watermarks.

Assigners would work just like any other operator. Assigners overriding the input watermark will lead to odd observations: having a source (that emits watermarks) followed by a watermark assigner may result in different output/input watermarks although they _should_ be identical (source emits 2L, but assigner "receives" 4L).

To _me_ this seems more intuitive, and it's definitely less complex :/

Reply:
Sure, I think that makes sense.